### DISCUSSION TITLE
Sharing a variable between jobs

### DISCUSSION THREAD

**Discussion by schof** [community/community#26313] [Actions]
<p>I have a common task that I need to perform for all job “variants” in a matrixed build. My idea was to use a separate job for that but I’m unsure how to pass a build output between two jobs.</p>
<p>Before needing to do a matrixed build (where we build both versions in parallel), I had been doing this all as one job. My question now is, what is the best way to share information from one job (the newly incremented verison) with the matrixed job that depends on it?</p>
<p>See below for some abberviated code. I haven’t tried to run it yet but I’m sure it will fail since AFAIK you can’t refernce steps from another job. I’m also aware of artifacts but we’re talking about a variable here and not some file.</p>
<p>TIA</p>
<pre><code>jobs:
  # increment version, tag and create a release
  release:
    name: Create Release
    ...
    steps:
      - name: Bump version
        uses: ...
        with:
          repo_token: ${{ secrets.GITHUB_TOKEN }}
          major_version: 62
        id: bump_version
  build:
    name: Build APK
    needs: release
    ...
    strategy:
      matrix:
        flavor: [sandbox,live]
    steps:
      - name: Checkout
        uses: actions/checkout@v1
      - name: Set version
        run: ./client/scripts/set-version.sh
        env:
          NEW_VERSION: ${{ steps.bump_version.outputs.new_version }}
        id: set_version
</code></pre>

**Comment by Yanjingzhu** [comment]
<p>As jobs run in different runners, those runners are in different machines. Each job runs in a fresh instance of the virtual environment specified by runs-on. It is not possible to share environment variables between machines.  We don’t support to share variables between jobs. <br>
As you metioned artifacts, we would suggest you use it . You can write the version into a file , upload it as artifact in one job and then download the artifacts in the other job , read the file content. </p>
<p>Please refer to: <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/persisting-workflow-data-using-artifacts" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/persisting-workflow-data-using-artifacts </a></p>

**Reply by eliezercazares** [reply]
how about environment variables?
those should be shared as they are higher in hierarchy in the yaml, than the jobs.

**Comment by schof** [comment]
<p>I hadn’t thought of writing the variable to a file but I guess that could work. Thanks for clarifying that it’s not possible to share information any other way. I do think this could be a feature worth adding at some point. Writing to a file seems like overkill for a single variable. Thanks again. </p>

**Comment by distinctlab** [comment]
<p>How to explain therefore that we can define environment variables at a global level as follows (thanks):</p>
<pre><code>name: Create release
 on:
  pull_request:
    types: closed
env:
  NODE_VERSION: "12.x"
  APP_VERSION: "1.0.O"
  ....
Jobs: ....
</code></pre>

**Comment by Yanjingzhu** [comment]
<p>Hi distinctlab, </p>
<p>Yeah, you could set a workflow level environment variable, which could be used in all jobs. But the original poster needs set the env variable inside a job. Then the env could only be used in that job.  </p>

**Comment by aroller** [comment]
<p>Thanks for the tip. I found Github <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts#passing-data-between-jobs-in-a-workflow" rel="nofollow noopener">documentation</a> provides how to share data between jobs. </p>
<pre><code>name: Share data between jobs

on: [push]

jobs:
  job_1:
    name: Add 3 and 7
    runs-on: ubuntu-latest
    steps:
      - shell: bash
        run: |
          expr 3 + 7 &gt; math-homework.txt
      - name: Upload math result for job 1
        uses: actions/upload-artifact@v1
        with:
          name: homework
          path: math-homework.txt

  job_2:
    name: Multiply by 9
    needs: job_1
    runs-on: windows-latest
    steps:
      - name: Download math result for job 1
        uses: actions/download-artifact@v1
        with:
          name: homework
      - shell: bash
        run: |
          value=`cat homework/math-homework.txt`
          expr $value \* 9 &gt; homework/math-homework.txt
      - name: Upload math result for job 2
        uses: actions/upload-artifact@v1
        with:
          name: homework
          path: homework/math-homework.txt

  job_3:
    name: Display results
    needs: job_2
    runs-on: macOS-latest
    steps:
      - name: Download math result for job 2
        uses: actions/download-artifact@v1
        with:
          name: homework
      - name: Print the final result
        shell: bash
        run: |
          value=`cat homework/math-homework.txt`
          echo The result is $value
</code></pre>

**Comment by distinctlab** [comment]
<pre><code> 
</code></pre>
<p> Thanks for your feedback. One of the current imits of github actions is to be able to declare from a job a global (workflow)<br>
variable that can be shared with other jobs.<br>
I’m talking about sharing variables not data (artifacts).<br>
Downloasd/upload artifacts actions is not practical to meet this need.</p>

**Comment by Yanjingzhu** [comment]
<p>Yeah, you are right. Github Actions could not declare a global env in one job and use it in other jobs in a workflow. I would recommend you to share your idea in the <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">Feedback form for GitHub Actions</a>.</p>

**Comment by asbjornu** [comment]
<p>In the same way it’s possible to share variables between steps, it should be possible to share variables between jobs as well. Since a job can declare its dependence on another job, the dependee could easily have its ouput variables persisted and exposed to dependent jobs through an expansions of the existing steps.&lt;step_id&gt;.outputs.&lt;output_name&gt; syntax with jobs.&lt;job_id&gt;.steps.&lt;step_id&gt;.outputs.&lt;output_name&gt;. Example below:</p>
<pre><code>jobs:
  version_job:
    name: Create version number
    runs-on: ubuntu-latest
    steps:
    - id: version_step
      run: echo "::set-output name=version_number::1.2.3"

  output:
    name: Output
    runs-on: ubuntu-latest
    needs: version_job
    steps:
    - run: echo "${{ jobs.version_job.steps.version_step.outputs.version_number }}"
</code></pre>

**Reply by NikolozCh** [reply]
(Un)fortunately, this feature is deprecated as of Mar. 2024.

**Comment by bradymholt** [comment]
<p>The new <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs" rel="nofollow noopener">jobs.&lt;jobs_id&gt;.outputs</a> syntax should support this.</p>
<pre><code>jobs:job1:runs-on:ubuntu-latest# Map a step output to a job outputoutputs:output1:${{steps.step1.outputs.test}}output2:${{steps.step2.outputs.test}}steps:-id:step1run:echo"::set-output name=test::hello"-id:step2run:echo"::set-output name=test::world"job2:runs-on:ubuntu-latestneeds:job1steps:-run:echo${{needs.job1.outputs.output1}}${{needs.job1.outputs.output2}}
</code></pre>

**Comment by cerna** [comment]
<p>Thanks for heads up, <a class="mention" href="/bradymholt">@bradymholt</a>, this is extremely helpful.</p>
<p>And exactly what I needed.</p>

**Comment by cryptable** [comment]
<p>Probably way too late, but I use this extension to share variables between jobs.</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/marketplace/actions/share-jobs-data" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/1122035/181096842-d1e3594a-a072-4b66-94c6-814bc984a937.jpeg" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/marketplace/actions/share-jobs-data" target="_blank" rel="nofollow noopener">Share Jobs Data - GitHub Marketplace</a></h3>

<p>Share data between GitHub Actions jobs</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I use it to release my artifacts into one release, so to store the release URL, which I need in another job.</p>
<pre><code class="lang-auto">...
    - name: set upload_url
      id: set_upload_url
      uses: noobly314/share-data@v1
      with:
        share-id: release_url
        mode: set
        key: url
        value: ${{ steps.create_release.outputs.upload_url }}
</code></pre>
<p>To get the value so I can use it again.</p>
<pre><code class="lang-auto">    - name: get upload_url
      id: get_upload_url
      uses: noobly314/share-data@v1
      with:
        share-id: release_url
        mode: get
        key: url
</code></pre>
<p>to access the data you do: ${{ steps.get_upload_url.outputs.data }}.</p>

**Comment by dimisjim** [comment, marked as answer]
**EDIT 2022/10**

Due to deprecation of `set-output` the reccomended approach is now: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter

```
echo "{name}={value}" >> $GITHUB_OUTPUT

      - name: Set color
        id: random-color-generator
        run: echo "SELECTED_COLOR=green" >> $GITHUB_OUTPUT
      - name: Get color
        run: echo "The selected color is ${{ steps.random-color-generator.outputs.SELECTED_COLOR }}"
```

**Reply by IanVS** [reply]
Does this share the output to other jobs?  The example does not indicate that it does.  This seems to be the real solution, defining an output map for the job itself: https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

**Reply by dimisjim** [reply]
yes, it does.

**Reply by kautlenbachs** [reply]
maybe I typoed, but I got it working by this -  https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

**Reply by SimonDarksideJ** [reply]
As a clarification on this answer, the above configuration will ONLY work on Linux distro's using Bash.

For Powershell you need to be more specific

```
echo "{name}={value}" >> $env:GITHUB_OUTPUT

      - name: Set color
        id: random-color-generator
        run: echo "SELECTED_COLOR=green" >> $env:GITHUB_OUTPUT
      - name: Get color
        run: echo "The selected color is ${{ steps.random-color-generator.outputs.SELECTED_COLOR }}"
```
**You HAVE to specify the output as an environment variable for Powershell or other scripting pipelines with the "env" prefix**

Hope this helps

**Comment by MostefaKamalLala** [comment]
<p>@ <a href="https://github.community/u/dimisjim">dimisjim</a> response should be the solution. Please change it just in case someone doesn’t scroll down enough!</p>

**Comment by m4heshd** [comment]
<p>@<a href="https://github.community/u/MostefaKamalLala">MostefaKamalLala</a> I flagged the current solution and got the staff to change the solution. :clap:t2:</p>

**Comment by SimonDarksideJ** [comment]
<p>Agreed, it’s a shame there is not a more flexible solution for sharing data between jobs, as even the parameters option has so many limitations (ran in to another one today)</p>
<p>In short:</p>
<ul>
<li>
<p>If it’s a single parameter, string (not JSON) or int, then you need to add boilerplate to surface outputs from steps to jobs, and then from Jobs to jobs.  However, a Job needs to “Depend” on another job to share the data (at least what I’ve found)</p>
</li>
<li>
<p>If it is more data than a single string / int, then you need to save that information to a file in an area sharable by different jobs, as each job can / will clear out the working folder between runs (or may be run on separate machines)</p>
</li>
<li>
<p>In reality, if there is a larger requirement for sharing data, you will need to invest time in some sort of state backend API that jobs can post and read data from.</p>
</li>
</ul>
<p>Here’s hoping for more improvement in this area.</p>

**Comment by ukreddy-erwin** [comment]
<p>I tried similar but not working for me.</p>
<pre><code class="lang-auto">jobs:
  # This workflow contains a single job called "build"
  build:
    # The type of runner that the job will run on
    runs-on: DevBuild1

    outputs:
      output1: ${{ steps.build_number.outputs.buildNumber }}
    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0
           
      # Runs a single command using the runners shell
      - name: git commit count
        run: |
          pwd
          echo "checking count"
          $count = git log --oneline | wc -l
          echo $count
          echo "git_commit_count=$count" &gt;&gt; $env:GITHUB_ENV
      - name: Print Build Number
        run: echo ${{ env.git_commit_count }}
      - name: export Build Number
        id: build_number
        run: echo "::set-output name=buildNumber::${{ env.git_commit_count }}"

 createInstaller:
    # The type of runner that the job will run on
    runs-on: DevBuild2
    needs:
      - build
    env:
      git_commit_count: ${{needs.build.outputs.buildNumber}}
    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      - name: Print build number
        run: echo ${{ env.git_commit_count }}
</code></pre>
<p>It is failing as value not detected.</p>
<pre><code class="lang-auto">echo
MissingMandatoryParameter,Microsoft.PowerShell.Commands.WriteOutputCommand
</code></pre>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="ukreddy-erwin" data-post="19" data-topic="16967">
<div class="title">
<div class="quote-controls"></div>
 ukreddy-erwin:</div>
<blockquote>
<p><code>      git_commit_count: ${{needs.build.outputs.buildNumber}}</code></p>
</blockquote>
</aside>
<p>There’s no output called <code>buildNumber</code> in the <code>build</code> job, the one that exists is called <code>output1</code>.</p>

**Comment by alienxp03** [comment]
Unfortunately, using output is deprecated. It's being replaced by Environment Files. However, it doesn't look like it supports sharing env between jobs. Only sharing env between steps is supported right now. So it seems like we're back to square one again.

```
The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#environment-files

**Reply by airtower-luna** [reply]
That's incorrect. The `set-output` *command* is deprecated, but not outputs as such. You just have to set them by writing to the file given in the `GITHUB_OUTPUT` environment variable instead, see: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter

The format for that new file is indeed the same as for the one already used for setting environment variables.

**Reply by alienxp03** [reply]
Ah. That's true. Thanks. I missed that part it seems.

**Reply by OctavianCahyadi** [reply]
still cannot get GITHUB_OUTPUT env value at jobs level @airtower-luna 

```
build-image:
    runs-on: ubuntu-latest
    steps:
         //other steps
       - name: Build the Docker image
          id: image
          run: |
            export VERSION=$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)-dev
            echo "TAG=${VERSION}" >> $GITHUB_OUTPUT

 update-deployment:
    needs: build-image
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Update Development project using Webhook
        run: |
          export TAG=${{ needs.build-image.steps.image.outputs.TAG }}

```

**Reply by airtower-luna** [reply]
The *step* output from a previous job won't be available, you need to create a job output from the step output. See: https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

**Comment by drsm79** [comment]
Does this work in a matrix job? If I have a matrix the steps are the same, but run with different configurations. How does that appear if the output is "flat"?

**Comment by joshbondaval** [comment]
I couldn't get this to work if the job that you want to output is ran with `uses`.  Does anyone know if outputs are supported if the job that 
performs the output is declared with `uses` property?


For example:

- The first job runs with `uses` which is referencing a shared workflow template. This job declares the `outputs` as described [here](https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs).
- The second job runs with `steps` and cannot obtain the outputs from the first job.

Here is an example of the main file:
```
on:
  pull_request:
    branches:
      - main

jobs:
  // ...
  upload: // <-- this job has the output as declared in upload_main_revision.yaml
    uses: bondaval/.github/.github/workflows/upload_main_revision.yml@main
    secrets: inherit
    with:
      app_name: app

  version:
    runs-on: ubuntu-latest
    needs: upload
    steps:
      - name: test if we can see version
        env:
          VERSION: ${{ needs.upload.outputs.VERSION }}
        run: |
          echo "$VERSION" // <--- Does not output VERSION as set by the 'upload' job.
```

This is the file that delcares the steps in `upload` job.
```
// upload_main_revision.yml
on:
  workflow_call:
    inputs:
      app_name:
        required: true
        type: string

jobs:
  upload:
    runs-on: ubuntu-latest
    outputs:
      VERSION: ${{ steps.output-build-number.outputs.VERSION }}
    steps:
       // ...
      - name: Create build version // <--- creates the next build version.
        id: version
        uses: paulhatch/semantic-version@v4.0.3
        with:
          // ...

      - name: Set build version
        run: echo "VERSION=${{ steps.version.outputs.version }}" >> $GITHUB_ENV

      - name: Output build number
        id: output-build-number
        run: echo "VERSION=${{ env.VERSION }}" >> "$GITHUB_OUTPUT"

```

**Reply by pavelfomin** [reply]
> I couldn't get this to work if the job that you want to output is ran with `uses`. Does anyone know if outputs are supported if the job that performs the output is declared with `uses` property?
> 
> For example:
> 
> * The first job runs with `uses` which is referencing a shared workflow template. This job declares the `outputs` as described [here](https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs).
> * The second job runs with `steps` and cannot obtain the outputs from the first job.
> 
> Here is an example of the main file:
> 
> ```
> on:
>   pull_request:
>     branches:
>       - main
> 
> jobs:
>   // ...
>   upload: // <-- this job has the output as declared in upload_main_revision.yaml
>     uses: bondaval/.github/.github/workflows/upload_main_revision.yml@main
>     secrets: inherit
>     with:
>       app_name: app
> 
>   version:
>     runs-on: ubuntu-latest
>     needs: upload
>     steps:
>       - name: test if we can see version
>         env:
>           VERSION: ${{ needs.upload.outputs.VERSION }}
>         run: |
>           echo "$VERSION" // <--- Does not output VERSION as set by the 'upload' job.
> ```
> 
> This is the file that delcares the steps in `upload` job.
> 
> ```
> // upload_main_revision.yml
> on:
>   workflow_call:
>     inputs:
>       app_name:
>         required: true
>         type: string
> 
> jobs:
>   upload:
>     runs-on: ubuntu-latest
>     outputs:
>       VERSION: ${{ steps.output-build-number.outputs.VERSION }}
>     steps:
>        // ...
>       - name: Create build version // <--- creates the next build version.
>         id: version
>         uses: paulhatch/semantic-version@v4.0.3
>         with:
>           // ...
> 
>       - name: Set build version
>         run: echo "VERSION=${{ steps.version.outputs.version }}" >> $GITHUB_ENV
> 
>       - name: Output build number
>         id: output-build-number
>         run: echo "VERSION=${{ env.VERSION }}" >> "$GITHUB_OUTPUT"
> ```

you are missing:
```yml
on:
  workflow_dispatch:
  workflow_call:
    outputs:
      VERSION:
        value: ${{ jobs.upload.outputs.VERSION }}
```
